### PR TITLE
added unit tests  for pricing module controllers

### PR DESCRIPTION
**Pull request by Arunkumar0610**
added tests for #1037 #1069 #1076 
**SimplCommerce.Module.Pricing->Controllers - >CartRuleApiController & CartRuleUsageApiController**

### CODE REVIEW

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+            // Arrange
+            var cartRuleRepositoryMock = new Mock<IRepository<CartRule>>();
+            var controller = new CartRuleApiController(cartRuleRepositoryMock.Object);
+            var smartTableParam = new SmartTableParam()
+            {
+                Pagination = new Pagination() { Start = 1, TotalItemCount = 1, Number = 1, NumberOfPages = 1 },
+                Search = new Search() { },
+                Sort = new Sort() { }
+            };
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:36]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            var result = controller.List(smartTableParam);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:38]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+            Assert.IsType<JsonResult>(result);
+        }
+
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:41]:
```suggestion
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                EndOn = DateTime.Now.AddDays(7),
+                IsCouponRequired = true,
+                RuleToApply = "Coupon",
+                DiscountAmount = 10,
+                DiscountStep = 2,
+                MaxDiscountAmount = 50,
+                UsageLimitPerCoupon = 100,
+                UsageLimitPerCustomer = 1
+            };
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:64]:
```suggestion

            // Act
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            var result = await controller.Post(model);
+            // Assert
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:66]:
```suggestion

            // Assert
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            var createdAtActionResult = Assert.IsType<CreatedAtActionResult>(result);
+            Assert.Equal(nameof(CartRuleApiController.Get), createdAtActionResult.ActionName);
+            Assert.NotNull(createdAtActionResult.RouteValues["id"]);
+            cartRuleRepositoryMock.Verify(repo => repo.Add(It.IsAny<CartRule>()), Times.Once);
+            cartRuleRepositoryMock.Verify(repo => repo.SaveChangesAsync(), Times.Once);
+        }
+
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:73]:
```suggestion
```

**Review by hishamco — CHANGES_REQUESTED** [review]
Please format and remove white spaces

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            var controller = new CartRuleApiController(cartRuleRepositoryMock.Object);
+            var smartTableParam = new SmartTableParam()
+            {
+                Pagination = new Pagination() { Start = 1, TotalItemCount = 1, Number = 1, NumberOfPages = 1 },
+                Search = new Search() { },
+                Sort = new Sort() { }
+            };
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:38]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+                RuleToApply = "Coupon",
+                DiscountAmount = 10,
+                DiscountStep = 2,
+                MaxDiscountAmount = 50,
+                UsageLimitPerCoupon = 100,
+                UsageLimitPerCustomer = 1
+            };
+            
+
+            // Act
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:68]:
```suggestion
            // Act
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+using SimplCommerce.Module.Pricing.Areas.Pricing.Controllers;
+using SimplCommerce.Module.Pricing.Models;
+using Xunit;
+
+namespace SimplCommerce.Module.Pricing.Tests.Controllers
+{
+    public class CartRuleUsageApiControllerTests
+    {
+
+        [Fact]
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:21]:
```suggestion
        [Fact]
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+            var deserializeobj = JsonConvert.DeserializeObject<searchObject>(serializeobj);
+            Assert.Equal("TestRule", deserializeobj.Items[0].CartRuleName);
+            Assert.Equal("TestUser", deserializeobj.Items[0].FullName);
+            Assert.Equal("TestCoupon", deserializeobj.Items[0].CouponCode);
+            Assert.Equal(101, deserializeobj.Items[0].CartRuleId);
+            Assert.Equal(201, deserializeobj.Items[0].UserId);
+            Assert.Equal(301, deserializeobj.Items[0].OrderId);
+
+        }
+        public class Itemsobject
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:99]:
```suggestion

        public class Itemsobject
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            public string CartRuleName { get; set; }
+            public long UserId { get; set; }
+            public string FullName { get; set; }
+            public string CouponCode { get; set; }
+            public long OrderId { get; set; }
+            public DateTimeOffset CreatedOn { get; set; }
+
+
+        }
+        public class searchObject
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:112]:
```suggestion

        public class searchObject
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+        {
+            public List<Itemsobject> Items { get; set; }
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:114]:
```suggestion
            public List<Itemsobject> Items { get; set; }

```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+            public int NumberOfPages { get; set; }
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:115]:
```suggestion
            public int NumberOfPages { get; set; }

```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+            public int TotalRecord { get; set; }
+
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:117]:
```suggestion
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+        }
+
+
+    }
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:121]:
```suggestion
    }
```

File: test/SimplCommerce.Module.Pricing.Tests/SimplCommerce.Module.Pricing.Tests.csproj [hunk 16]
```diff
@@ -26,4 +26,8 @@
     <ProjectReference Include="..\..\src\SimplCommerce.Infrastructure\SimplCommerce.Infrastructure.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <Folder Include="Controllers\" />
+  </ItemGroup>
+
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/SimplCommerce.Module.Pricing.Tests.csproj:32]:
```suggestion
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+using Microsoft.EntityFrameworkCore.Query;
+using Moq;
+using SimplCommerce.Infrastructure.Data;
+using SimplCommerce.Infrastructure.Web.SmartTable;
+using SimplCommerce.Module.Catalog.Models;
+using SimplCommerce.Module.Pricing.Areas.Pricing.Controllers;
+using SimplCommerce.Module.Pricing.Areas.Pricing.ViewModels;
+using SimplCommerce.Module.Pricing.Models;
+using Xunit;
+namespace SimplCommerce.Module.Pricing.Tests.Controllers
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleApiControllerTests.cs:20]:
```suggestion

namespace SimplCommerce.Module.Pricing.Tests.Controllers
```

File: test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+            public long CartRuleId { get; set; }
+            public string CartRuleName { get; set; }
+            public long UserId { get; set; }
+            public string FullName { get; set; }
+            public string CouponCode { get; set; }
+            public long OrderId { get; set; }
+            public DateTimeOffset CreatedOn { get; set; }
+
+
+        }
```
Comment (hishamco) [inline test/SimplCommerce.Module.Pricing.Tests/Controllers/CartRuleUsageApiControllerTests.cs:111]:
```suggestion
        }
```

**Review by hishamco — CHANGES_REQUESTED** [review]

**Comment by hishamco** [comment]
@Arunkumar0610 please run the tests at your end and make sure everything works as expected

**Comment by Arunkumar0610** [comment]
@hishamco Yes I ran the tests at my end everything works as excepted

**Comment by Arunkumar0610** [comment]
@hishamco resolved the conflict and its merged successfully

**Pull request merged**